### PR TITLE
[GoRouter] option to override initial route set by platform 

### PR DESCRIPTION
**Pull request by opxdelwin**
Added optional parameter to override platform route mentioned in `MainActivity.kt`.

This would be completely optional and caters to a very edge-case, but providing an option to would be best. 

This PR will fix https://github.com/flutter/flutter/issues/132402 & Feature Req flutter/flutter#132557

All existing tests pass, no breaking changes.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/go_router/lib/src/router.dart [hunk 1]
```diff
@@ -509,6 +522,11 @@ class GoRouter implements RouterConfig<RouteMatchList> {
   String _effectiveInitialLocation(String? initialLocation) {
     final String platformDefault =
         WidgetsBinding.instance.platformDispatcher.defaultRouteName;
+    if (overridePlatformDefaultLocation) {
+      ///can force null check as it's already verified by
```
Comment (chunhtai) [inline packages/go_router/lib/src/router.dart:526]: code comment should be double slash, triple slash is reserved for doc string

File: packages/go_router/test/go_router_test.dart [hunk 2]
```diff
@@ -4863,6 +4863,29 @@ void main() {
       expect(statefulWidgetKeyB.currentState?.counter, equals(1));
     });
   });
+
+  ///Regression tests for https://github.com/flutter/flutter/issues/132557
+  group('overridePlatformDefaultLocation', () {
+    test('No initial location provided', () {
```
Comment (chunhtai) [inline packages/go_router/test/go_router_test.dart:4869]: This test only test for the assert, can you write a test that mocks the initial route from platform?

Comment (opxdelwin) [reply packages/go_router/test/go_router_test.dart:4869]: I will be writing test to emulate initial route from platform, but I'd probably need help to implement it. I'll keep you posted.

Comment (opxdelwin) [reply packages/go_router/test/go_router_test.dart:4869]: @chunhtai I was unable to find a way to set platform initalroute pragmatically for testing. Requesting help...

**Comment by opxdelwin** [comment]
@chunhtai any updates on this?

**Comment by chunhtai** [comment]
sorry I will give an example for test today

File: packages/go_router/test/go_router_test.dart (hunk 2 above)
Comment (chunhtai) [reply packages/go_router/test/go_router_test.dart:4869]: looks like there isn't an easy way. I think you can test it like this.
```dart
final String platformRoute = WidgetsBinding.instance.platformDispatcher.defaultRouteName;
final String expectedInitialRoute = '/abc';
// Sanity check.
expect(platformRoute != expectedInitialRoute, isTrue)
final GoRouter router = GoRouter(overridePlatformDefaultLocation: true, initialRoute: '/abc', ...);
expect(router.routeInformationProvider.value.location,  expectedInitialRoute);
```

Comment (chunhtai) [reply packages/go_router/test/go_router_test.dart:4869]: This seems not addressed yet?

**Review by chunhtai — COMMENTED** [review]
See https://github.com/flutter/packages/pull/4717/files#r1327848920

File: packages/go_router/test/go_router_test.dart (hunk 2 above)
Comment (opxdelwin) [reply packages/go_router/test/go_router_test.dart:4916]: Added new commits.

**Comment by opxdelwin** [comment]
Do review

File: packages/go_router/lib/src/router.dart [hunk 3]
```diff
@@ -299,6 +302,15 @@ class GoRouter implements RouterConfig<RouteMatchList> {
   @override
   late final GoRouteInformationParser routeInformationParser;
 
+  /// When set to [true], the route returned by `getInitialRoute` will take precedence over the
```
Comment (chunhtai) [inline packages/go_router/lib/src/router.dart:305]: There is no `getInitialRoute` in the code

File: packages/go_router/lib/src/router.dart [hunk 4]
```diff
@@ -299,6 +302,15 @@ class GoRouter implements RouterConfig<RouteMatchList> {
...
+  /// platform's default initial location. This allows developers to control the starting route
+  /// of the application independently of the platform.
+  ///
+  /// Setting this parameter to [false] (default) will allow the platform's default initial
+  /// location to be used. It's advisable to only set this to true if you explicitly want to
```
Comment (chunhtai) [inline packages/go_router/lib/src/router.dart:310]:
```suggestion
  /// location to be used even if the `initialLocation` is set. It's advisable to only set this to true if one explicitly want to
```

Comment (chunhtai) [reply packages/go_router/lib/src/router.dart:310]: our style guide suggest doc string should avoid using `I`, `we`, `you`.

File: packages/go_router/lib/src/router.dart (hunk 3 above)
Comment (chunhtai) [reply packages/go_router/lib/src/router.dart:305]: Also the first paragraph should be a brief description of what this property is

Comment (chunhtai) [reply packages/go_router/lib/src/router.dart:305]: e.g.

```
Whether to ignore platform's default initial location when `initialLocation` is set.
```

File: packages/go_router/lib/src/router.dart [hunk 5]
```diff
@@ -299,6 +302,15 @@ class GoRouter implements RouterConfig<RouteMatchList> {
   @override
   late final GoRouteInformationParser routeInformationParser;
 
+  /// When set to [true], the route returned by `getInitialRoute` will take precedence over the
+  /// platform's default initial location. This allows developers to control the starting route
+  /// of the application independently of the platform.
+  ///
```
Comment (chunhtai) [inline packages/go_router/lib/src/router.dart:308]: somewhere in the doc should talk about when platform's default initial location will be set. e.g. deeplink

File: packages/go_router/lib/src/router.dart [hunk 6]
```diff
@@ -509,6 +521,11 @@ class GoRouter implements RouterConfig<RouteMatchList> {
   String _effectiveInitialLocation(String? initialLocation) {
     final String platformDefault =
         WidgetsBinding.instance.platformDispatcher.defaultRouteName;
+    if (overridePlatformDefaultLocation) {
```
Comment (chunhtai) [inline packages/go_router/lib/src/router.dart:524]: should this be move to beginning of the method?

File: packages/go_router/test/go_router_test.dart [hunk 7]
```diff
@@ -4909,6 +4909,54 @@ void main() {
...
+                    path: '/b',
+                    builder: (BuildContext context, GoRouterState state) =>
+                        const Placeholder(),
+                  ),
+                ],
+              ),
+          throwsA(const TypeMatcher<AssertionError>()));
+    });
+    test('Test override using routeInformationProvider', () {
+      final String platformRoute =
```
Comment (chunhtai) [inline packages/go_router/test/go_router_test.dart:4935]: Sorry I wasn't thinking clearly, I realize the initialRoute will take priority if WidgetsBinding.instance.platformDispatcher.defaultRouteName is `/`.

Can you convert this test to `testWidget`

and use
```
tester.binding.platformDispatcher.defaultRouteNameTestValue('/some-route')
```
This should mock the platform default initial location. and then verify the `GoRouter(initialLocation:)` really take priority

File: packages/go_router/lib/src/router.dart (hunk 6 above)
Comment (opxdelwin) [reply packages/go_router/lib/src/router.dart:524]: cant as the underlying if conditions depend on `platformDefault`

File: packages/go_router/test/go_router_test.dart (hunk 7 above)
Comment (opxdelwin) [reply packages/go_router/test/go_router_test.dart:4984]: Implemented

File: packages/go_router/lib/src/router.dart (hunk 5 above)
Comment (opxdelwin) [reply packages/go_router/lib/src/router.dart:307]: added these docs

File: packages/go_router/lib/src/router.dart (hunk 4 above)
Comment (opxdelwin) [reply packages/go_router/lib/src/router.dart]: I went over that, it's fixed

File: packages/go_router/lib/src/router.dart (hunk 6 above)
Comment (chunhtai) [reply packages/go_router/lib/src/router.dart:532]: I am not sure if I understand, the if condition returns initialLocation directly, it doesn't use platformDefault

Comment (opxdelwin) [reply packages/go_router/lib/src/router.dart:532]: `platformDefault` would send out a string:
 - `'/'`                    if no route is provided by platform on startup (eg. by deeplink)
 - `'/some-route'` if specific route is provided by platform on startup (eg. by deeplink)
 
 if we're using `overridePlatformDefaultLocation`, we don't care about the `platformDefault` in the first place and return `initialLocation` set by router.

Feel free to crosscheck
<details>
<summary> image </summary>

image
</details>

File: packages/go_router/lib/src/router.dart [hunk 8]
```diff
@@ -509,6 +529,11 @@ class GoRouter implements RouterConfig<RouteMatchList> {
...
+      // can force null check as it's already verified by
+      // assert() while initialization
+      return initialLocation!;
+    }
```
Comment (chunhtai) [inline packages/go_router/lib/src/router.dart:536]:
```suggestion
    if (overridePlatformDefaultLocation) {
      // can force null check as it's already verified by
      // assert() while initialization
      return initialLocation!;
    }
    final String platformDefault =
        WidgetsBinding.instance.platformDispatcher.defaultRouteName;
```

**Review by chunhtai — COMMENTED** [review]
wouldn't this work?

File: packages/go_router/lib/src/router.dart (hunk 8 above)
Comment (chunhtai) [reply packages/go_router/lib/src/router.dart:536]: The idea is to avoid declare the `platformDefault` if `overridePlatformDefaultLocation` is true.

Comment (opxdelwin) [reply packages/go_router/lib/src/router.dart:536]: oh right, it'd definitely work! I'll commit, thanks.

File: packages/go_router/lib/src/router.dart [hunk 9]
```diff
@@ -299,6 +302,23 @@ class GoRouter implements RouterConfig<RouteMatchList> {
   @override
   late final GoRouteInformationParser routeInformationParser;
 
+  ///Whether to ignore platform's default initial location when
+  ///`initialLocation` is set.
```
Comment (chunhtai) [inline packages/go_router/lib/src/router.dart:306]:
```suggestion
  /// Whether to ignore platform's default initial location when
  /// `initialLocation` is set.
```

File: packages/go_router/lib/src/router.dart [hunk 10]
```diff
@@ -299,6 +302,23 @@ class GoRouter implements RouterConfig<RouteMatchList> {
...
+  ///
+  /// When set to [true], the route set as [initialLocation] will take
```
Comment (chunhtai) [inline packages/go_router/lib/src/router.dart:308]:
```suggestion
  /// When set to [true], the [initialLocation] will take
```

File: packages/go_router/lib/src/router.dart [hunk 11]
```diff
@@ -507,6 +527,11 @@ class GoRouter implements RouterConfig<RouteMatchList> {
   }
 
   String _effectiveInitialLocation(String? initialLocation) {
+    if (overridePlatformDefaultLocation) {
+      // can force null check as it's already verified by
+      // assert() while initialization
```
Comment (chunhtai) [inline packages/go_router/lib/src/router.dart:532]:
```suggestion
      // The initialLocation must not be null as it's already 
      // verified by assert() during the initialization.
```

**Review by chunhtai — APPROVED** [review]
LGTM

**Comment by chunhtai** [comment]
looks like the ci is failing. can you fix it?

**Comment by opxdelwin** [comment]
@chunhtai all green, https://github.com/flutter/packages/labels/autosubmit may be added manually

**Comment by opxdelwin** [comment]
awaiting review @hangyujin

**Review by hangyujin — APPROVED** [review]
LGTM

**Comment by chunhtai** [comment]
The failures are not related to this pr, need to wait for main branch to resolve the failures before this can merge

**Comment by opxdelwin** [comment]
Noted, we'll let the PR merge itself.
